### PR TITLE
docs: update storage architecture diagram

### PR DESCRIPTION
**Pull request by galexrt**
**Description of your changes:**

This includes the source code of the diagram, created using draw.io, so anyone can easily update it.

It has slightly less "detail," but we can edit it now and keep it updated for future components that should be mentioned on the storage architecture.

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [x] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by travisn — CHANGES_REQUESTED** [review]
A few thoughts on the diagram...
- CSI driver pods can be added for rbd and cephfs provisioners and volume plugins
- I'd rather not show discovery in the main diagram since it's disabled by default and not needed very often
- Let's show balanced nodes with the same number of OSDs. For example, how about three OSDs on each of three nodes
- There are other daemons such as crash collector and exporter. We might consider adding them, but I'm thinking not so we can just focus on the core components. 
- For consistency, can the apps have rounded corners like the ceph components?
- What is the green app? Is it not using the ceph storage? I like the idea of having three apps, which each use rbd (RWO), cephfs (RWX), and rgw (S3). And to show cephfs sharing, perhaps we actually have two apps that share the same pvc, so we would have four total apps.
- Since this will be on the architecture doc page, we can remove the "Rook Architecture" title

**Review by BlaineEXE — COMMENTED** [review]
The way the OSDs are depicted could lead someone to assume all OSDs are run within a single pod. Perhaps they should be separated.

Also, while Ceph *can* support clients running with different numbers of OSDs, it doesn't prefer that. Should we be suggesting to users that they shouldn't *try* to have homogeneous nodes?

I don't see that nodes are delineated. Someone might assume all of these are running on a single node.

Should we include CSI in this diagram? CephFS and RBD seem like they'd be good to include. Additionally, NFS is available, though disabled by default.

What does the dotted line represent? Is that a network? It should be labeled.

It seems strange that Rook/Ceph pods are round-edged but user app pods are square-edged.

Is it important to call out RWX (read-write many) for a multi-replica app requesting CephFS storage by having a scale-out user app in the diagram? I'm ambivalent, but it would be good to consider.

I see "<logo> Kubernetes", but it doesn't have a clear indication of its scope. What is inside that k8s cluster?



In general, I think it would be good to see more labels on what things are: nodes, networks, pods, cluster, should be clear for readers. 

K8s has an icon library that could be helpful : https://github.com/kubernetes/community/blob/master/icons/README.md

I almost wonder about just avoiding OBC altogether and just starting with COSI BucketAccessRequest resources. Especially if we wait to release this diagram with Rook v1.12, it will also be accurate, and implicitly begins the deprecation of OBCs by de-emphasizing it in favor of its long-term replacement.

**Comment by BlaineEXE** [comment]
> Since this will be on the architecture doc page, we can remove the "Rook Architecture" title

I would provide a contrasting opinion that including "Rook Architecture" could be good for any user that wants to copy-paste it somewhere. I'm a proponent of things having self-evident context, and in a contextless environment like a bing image search, the text in the image will help provide that context. And if I want to use it for a presentation, my old friend `crop and zoom tool` can do the necessary to shear the title.

**Comment by travisn** [comment]
> > Since this will be on the architecture doc page, we can remove the "Rook Architecture" title
> 
> I would provide a contrasting opinion that including "Rook Architecture" could be good for any user that wants to copy-paste it somewhere. I'm a proponent of things having self-evident context, and in a contextless environment like a bing image search, the text in the image will help provide that context. And if I want to use it for a presentation, my old friend `crop and zoom tool` can do the necessary to shear the title.

True, sounds good to keep the title on the image in case it's used in another context

**Comment by galexrt** [comment]
I have updated it using the Kubernetes Icons Set and tweaked the structure based on the other feedback points:

image

Any thoughts? Does this address your points/ concerns?

**Comment by travisn** [comment]
A few more thoughts:
- Even with the large diagram, the "pod", "pv", "node" are very small and difficult to read on the diagram. Instead, what if we used color coding for the different types of resources? For example, namespaces, pods, and nodes would have a different color outline.
- OSDs could still use their own box instead of three in the same box
- Instead of CSI Driver, how about CSI Volume Plugin? Then users will know which component they are looking at when they install, since we're really showing pods.
- RGW S3 API: It looks like there is a ceph component in the client. Let's connect this to the rgw component in the ceph diagram below. The connection can mention S3 API, but it's confusing with ceph in that client diagram.
- COSI isn't released yet. How about we don't mention COSI or OBC, it's just an S3 client. I don't think this overall arch diagram needs COSI components even after we release.
- In the application layer, I'm not sure we need the details for PVC, PV, and SC. If we can cut those out it will simplify the diagram and keep the focus on rook arch. 
- Is there a different disk icon to include with OSDs? Maybe it's just me, but the icon is not obviously a disk.

**Comment by travisn** [comment]
Before we finalize the diagram, as we discussed in huddle, let's decide what diagrams we want to create. This will help us decide the level of detail to include in each diagram. We just can't get all the detail in a single diagram or else it is too overwhelming, especially for new users just learning about Rook. All the useful detail you have in this diagram is good, let's just figure out how to split it up.

The first diagram on the arch page must be high level and simple for new users. The main diagram we currently have feels like the right level of detail, but it just needs to be updated. For example:
    a. We need to clearly show that we support three types of storage. The first app should show a "RWO Volume Claim", the second app should show a "RWX Volume Claim", and the third app should show that it is using an S3 client to connect to object storage. There should not be a fourth app.
    b. Clearly show color coded boxes for three different types of pods: Rook pod (operator: blue with a Rook logo), CSI pods (Provisioner and Volume Plugin: orange with the CSI logo), then Ceph pods (mon, OSD, MDS, RGW: red with a ceph logo). There is no need to show both rbd and cephfs drivers, we can just combine them in this high level diagram.

Then we have three more detailed diagrams for each type of storage: block, shared filesystem, and object. Each of them could show the details with PVC, PV, storage class, specific rbd/cephfs drivers, etc. Perhaps these diagrams could just lump all the Ceph pods into a single box to keep that layer simple and focus on the application usage. Or in some cases it will help to show the Ceph pod, specifically RGW or MDS.

**Comment by galexrt** [comment]
I would currently propose the following as the "new" high level architecture diagram
image

Then based on @travisn feedback, the rook-ceph namespace architecture for "Detailed view of a Rook Ceph Cluster" and then per storage type the diagrams using the Kubernetes icons and arrows to each "box" of a component (e.g., Ceph cluster is just one box for simplicity).

What are your thoughts?

**Comment by travisn** [comment]
Looking great, just a few more thoughts on some finer details:
- Koor title? :) 
- OSDs could still use their own box instead of three in the same box
- Instead of CSI "Driver", how about "Volume Plugin"? Then users will know which component they are looking at when they install to correspond to pods. There is no pod called "Driver".
- Can you round the corners of the applications, similar to the rook/ceph pods?
- Instead of showing the PVCs and S3 client as separate resources, how about having them inside the same App rectangle? Since the volumes are mounted inside the app pod, and the s3 client is used by the same app pod, this makes sense to me. Although the two lines to the RWX PVC do make an interesting point about multiple apps, even if multiple RWX apps are not shown. At a minimum, for consistency perhaps use a rectangle instead of oval for the S3 API. 
- How about a vertical dotted line to separate nodes in the lower diagram? Otherwise, I'm not sure it's clear that we're attempting to represent three separate nodes.

**Comment by galexrt** [comment]
> - Instead of showing the PVCs and S3 client as separate resources, how about having them inside the same App rectangle? Since the volumes are mounted inside the app pod, and the s3 client is used by the same app pod, this makes sense to me. Although the two lines to the RWX PVC do make an interesting point about multiple apps, even if multiple RWX apps are not shown. At a minimum, for consistency perhaps use a rectangle instead of oval for the S3 API. 

How about changing "App" to "Pod" and making it two Pods for the RWX volume?

***

image

**Comment by parth-gr** [comment]
Symbol should be rook operator for more clarity

**Comment by travisn** [comment]
This diagram looks good to me. Just one more nit, is that it would look nicer if the apps are the same width as the PVC or S3 box.

**Comment by travisn** [comment]
The main architecture image looks good. The image is just rather large, can you shrink the resolution?
I'll take a look at the new images a bit later...

File: Documentation/Getting-Started/ceph-storage/architecture.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/architecture.png:1]: This resolution still looks very large. If you render the docs with this diagram, does it automatically shrink to the width of the doc? Maybe it will be fine in that case.

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png:1]: - What are "bucket details"?
- "cm": What if we leave off this component and just describe it as a written detail as part of "bucket credentials"
- What is above the pods? Is that a "node"? It's too small to read. I think the dotted box could be naturally be understand as nodes. Alternatively, maybe we don't need the node boxes at all. It's not an important detail IMO that the pods are on different nodes. They might be on the same nodes as each other, or the ceph pods.
- "Ceph Cluster": How about naming this "Ceph RGW"? Then in the diagram description, we can mention all ceph components back the RGW daemon.
- Why have two application pods? One seems sufficient.
- Instead of a pod, what if we have an orange "App" so they can relate back to the overall diagram.
- What about adding a "Bucket" resource underneath the "Ceph RGW"? Then they can see where the data is stored.

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png:1]: - Can we have bigger labels and spell out sc, pv, and pvc? While common in K8s, newcomers won't know what they are. Alternatively, we could define them in text below the diagram.
- "Ceph Cluster" --> "Ceph RBD"
- Instead of a pod, how about a blue "App" to relate back to the main diagram?
- "Mounting" --> "Mount"
- No need for "node" icon, the dotted line seems sufficient, and then we can explain below the diagram.
- Center the title on the diagram
- For the title: "RWO Volume". They can see the main diagram for the meaning of RWO, or also in the description below the diagram. It's nice to have concise titles.
- "Provisioner" --> "RBD Provisioner"
- "CSI Driver" --> "CSI RBD Plugin"

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWX.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWX.png:1]: - Instead of a pod, how about a purple "App"
- "Provisioner" --> "CephFS Provisioner"
- Title: "RWX Volume"
- "Ceph Cluster" --> "Ceph MDS"
- "Mounting" --> "Mount"
- "CSI Driver" --> "CSI CephFS Plugin"

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:38]: Below each of these diagrams it would be great to have a description of them, and each component. I'm happy to  help write these descriptions after we have finalized the diagrams.

**Review by travisn — CHANGES_REQUESTED** [review]

File: Documentation/Getting-Started/ceph-storage/architecture.png
Comment (galexrt) [reply Documentation/Getting-Started/ceph-storage/architecture.png:1]: Yes, the image is constrained to the width of the documentation page.

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWX.png
Comment (galexrt) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWX.png:1]: I would prefer keeping the title as the long form with the short form in brackets. Not everyone might know what RWX or RWO means, so it feels like having both makes it a good compromise

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png
Comment (galexrt) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png:1]: > No need for "node" icon, the dotted line seems sufficient, and then we can explain below the diagram.

For clarity of people not knowing the "limitations" of RBD/ CephFS, I would want to keep the node label in, so anyone knows it is a node.

I have changed the name to be in line but also point out what the short `RWO` means.

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWX.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWX.png:1]: Instead of the node icon, how about just the word "node"? The dotted line represents the node so I don't think the icon is necessary for the node.

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png:1]: How about if the legends are not part of the image? I was thinking a text paragraph could be shown above or below the image on the help page. Then we can explain it more.

**Review by travisn — CHANGES_REQUESTED** [review]
The diagrams are looking great, just a couple more suggestions...

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png
Comment (galexrt) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png:1]: I have moved them to the bottom left in the diagrams now

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:40]: Description paragraphs needed

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png:1]: How about if the legend has smaller font like in the other diagrams? It's nice if the legend is a small detail to be viewed instead of such a big box.

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png:1]: I still think the node icon is distracting and would prefer just the word "node" on top of the dotted line, if possible.

**Review by travisn — CHANGES_REQUESTED** [review]

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png
Comment (galexrt) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png:1]: @travisn though the clear K8S coloring of the "node" icon, indicates that it is "part" of Kubernetes.  replacing it by just a "node" as a text?

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png
Comment (galexrt) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png:1]: I have set the same font size in the 3 diagrams now.

File: Documentation/Getting-Started/storage-architecture.md
Comment (galexrt) [reply Documentation/Getting-Started/storage-architecture.md:40]: Working on them tomorrow

Comment (travisn) [reply Documentation/Getting-Started/storage-architecture.md:40]: I'm sure I'll have input :), thanks for putting together the draft!

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png
Comment (travisn) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png:1]: hmm it doesn't look the same size, it must just be how the images are scaled in this view

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png
Comment (travisn) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png:1]: To me, the representation of the node is the dotted line. At first glance the icon looks like a separate resource, so I think it will be simpler for the user to understand if there is not an icon in addition to the dotted line. Maybe that's just me.

Comment (galexrt) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png:1]: Well.. technically a Node is a resource in Kubernetes :smile: :wink:

@BlaineEXE @parth-gr any thoughts from your side regarding the node icon vs node text/label?

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:36]:
```suggestion
### Block Storage

![Rook Volume ReadWriteOnce Diagram](ceph-storage/Rook%20-%20Storage%20Type%20Volume%20RWO.png)

In the diagram above, the flow to create an application with an RWO volume is:
- The (blue) app creates a PVC to request storage
- The PVC defines the Ceph RBD storage class (sc) for provisioning the storage
- K8s calls the Ceph-CSI RBD provisioner to create the Ceph RBD image. 
- The kubelet calls the CSI RBD volume plugin to mount the volume in the app
- The volume is now available for reads and writes.
```

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:44]:
```suggestion
### Shared Filesystem

![Rook Volume ReadWriteMany Diagram](ceph-storage/Rook%20-%20Storage%20Type%20Volume%20RWX.png)

In the diagram above, the flow to create a applications with a RWX volume is:
- The (purple) app creates a PVC to request storage
- The PVC defines the CephFS storage class (sc) for provisioning the storage
- K8s calls the Ceph-CSI CephFS provisioner to create the CephFS subvolume
- The kubelet calls the CSI CephFS volume plugin to mount the volume in the app
- The volume is now available for reads and writes.
```

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png
Comment (travisn) [inline Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png:1]: How about adding an "S3 Client" box between the app and the Ceph RGW component? Then it will match the arch diagram, and the description.

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:52]:
```suggestion
### Object Storage

![Rook Object Storage S3 Diagram](ceph-storage/Rook%20-%20Storage%20Type%20S3.png)

In the diagram above, the flow to create an application with access to an S3 bucket is:
- The (orange) app creates an ObjectBucketClaim (OBC) to request a bucket
- The Rook operator creates a Ceph RGW bucket (via the lib-bucket-provisioner)
- The Rook operator creates a secret with the credentials for accessing the bucket and a configmap with bucket information
- The app retrieves the credentials from the secret
- The app can now read and write to the bucket with an S3 client
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: Documentation/Getting-Started/storage-architecture.md
Comment (dave-at-koor) [reply Documentation/Getting-Started/storage-architecture.md:25]: Can the vertical lines separating the nodes be bolder? Or have three separate boxes, each with the layers? 

Does the operator run in one of the nodes? 

What does the dotted line represent? Is it the boundary to K8s? Aren't the apps also in K8s?

**Review by dave-at-koor — COMMENTED** [review]
Looks like a big improvement. Helps to have separate diagrams to show each kind of storage.

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png
Comment (travisn) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type S3.png:1]: Actually instead of a box, how about just labeling that line as "S3 Client". The box makes it feel like there is a separate pod or component. Thanks!

File: Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png
Comment (galexrt) [reply Documentation/Getting-Started/ceph-storage/Rook - Storage Type Volume RWO.png:1]: > @galexrt can you please show the latest image, and exactly where this `node` is pointing?

The latest images are in the PR, Travis is talking about the Kubernetes node icon:

image

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:76]:
```suggestion
### Object Storage S3
```

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [reply Documentation/Getting-Started/storage-architecture.md:40]: Sound good to move these two paragraphs up?

**Review by travisn — COMMENTED** [review]
@BlaineEXE @dave-at-koor @galexrt What are the remaining blocking issues we need to resolve before merging this? It would be nice to get this merged soon and hold off other details for future PRs. 

I would still propose the Node icon be changed to the word "Node" on the dotted line, I thought there was agreement when that came up in a recent huddle, though perhaps you weren't there @galexrt.

File: Documentation/Getting-Started/storage-architecture.md
Comment (galexrt) [reply Documentation/Getting-Started/storage-architecture.md:40]: @travisn Could you check the file again the line starting with "The `rook/ceph` image includes all necessary..." is a lot further down on my end. If you could use the suggestion feature to show which paragraphs you mean that would help

Comment (travisn) [reply Documentation/Getting-Started/storage-architecture.md:40]: I'm suggesting to move these middle two paragraphs after this first line and before the first architecture image. Then everything after the images is just referring to the images.
```suggestion
Rook automatically configures the Ceph-CSI driver to mount the storage to your pods.

The `rook/ceph` image includes all necessary tools to manage the cluster. Rook is not in the Ceph data path.
Many of the Ceph concepts like placement groups and crush maps
are hidden so you don't have to worry about them. Instead, Rook creates a simplified user experience for admins that is in terms
of physical resources, pools, volumes, filesystems, and buckets. Advanced configuration can be applied when needed with the Ceph tools.

Rook is implemented in golang. Ceph is implemented in C++ where the data path is highly optimized. We believe
this combination offers the best of both worlds.

![Rook Components on Kubernetes](ceph-storage/Rook High-Level Architecture.png)
```

**Review by parth-gr — COMMENTED** [review]
For Rook High level Architecture,

Do `mgr` is good with keeping with RBD node? 

Like it can be land up in any of the node, 
But it in above case it more generally looks like that for rbd the mgr pod is needed

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:40]: Shouldn't this Diagrams title be down on line 52? These paragraphs before that aren't related to the diagrams

Comment (galexrt) [reply Documentation/Getting-Started/storage-architecture.md]: @travisn Like this? Might have miscopied while I tried to fix the paragraphs move

Comment (travisn) [reply Documentation/Getting-Started/storage-architecture.md]: Looks good now, thanks

**Comment by galexrt** [comment]
I'll update the PR tomorrow as we have discussed leaving the storage type diagrams out of this PR and adding them in a separate PR. The paragraphs per storage type will be included in this PR.

**Comment by galexrt** [comment]
@travisn @BlaineEXE I have updated the PR to only include the new high level architecture diagram and the storage type descriptions. PTAL

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:24]: Lines 24-40 need to be removed. They are duplicated from the new diagram section.

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:49]: What if we remove the "Diagrams" title? They will obviously see the diagrams in the next sections anyway. 
```suggestion
```

**Review by travisn — CHANGES_REQUESTED** [review]

**Review by BlaineEXE — APPROVED** [review]

**Comment by galexrt** [comment]
@travisn PTAL

File: Documentation/Getting-Started/storage-architecture.md
Comment (travisn) [inline Documentation/Getting-Started/storage-architecture.md:24]: Need to remove this line, it is still duplicated from line 22
```suggestion
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: Documentation/Getting-Started/storage-architecture.md
Comment (galexrt) [reply Documentation/Getting-Started/storage-architecture.md:24]: @travisn Updated, PTAL

**Review by travisn — APPROVED** [review]
LGTM!

**Pull request merged**
